sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,411 @@
+        """
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss = self._loss.loss(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if w.ndim == 1:
+            return loss + 0.5 * l2_reg_strength * (w @ w)
+        else:
+            return loss + 0.5 * l2_reg_strength * squared_norm(w)
+
+    def loss_gradient(
+        self, coef, X, y, sample_weight=None, l2_reg_strength=0.0, n_threads=1
+    ):
+        """Computes the sum/average of loss and gradient.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,), (n_classes, n_dof) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+            If shape (n_classes * n_dof,), the classes of one feature are contiguous,
+            i.e. one reconstructs the 2d-array via
+            coef.reshape((n_classes, -1), order="F").
+        y : contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or contiguous array of shape (n_samples,)
+            Sample weights.
+        l2_reg_strength: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+
+        gradient : ndarray of shape coef.shape
+             The gradient of the loss.
+        """
+        n_features, n_classes = X.shape[1], self._loss.n_classes
+        n_dof = n_features + self.fit_intercept
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss, gradient_per_sample = self._loss.loss_gradient(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if not self._loss.is_multiclass:
+            loss += 0.5 * l2_reg_strength * (w @ w)
+            grad = np.empty_like(coef, dtype=X.dtype)
+            grad[:n_features] = X.T @ gradient_per_sample + l2_reg_strength * w
+            if self.fit_intercept:
+                grad[-1] = gradient_per_sample.sum()
+            return loss, grad
+        else:
+            loss += 0.5 * l2_reg_strength * squared_norm(w)
+            grad = np.empty((n_classes, n_dof), dtype=X.dtype, order="F")
+            # gradient.shape = (n_samples, n_classes)
+            grad[:, :n_features] = gradient_per_sample.T @ X + l2_reg_strength * w
+            if self.fit_intercept:
+                grad[:, -1] = gradient_per_sample.sum(axis=0)
+            if coef.ndim == 1:
+                return loss, grad.ravel(order="F")
+            else:
+                return loss, grad
```

Comment:
Looking at the code of this function, I am pretty sure we could rewrite it in Cython with a prange loop on chunks of X to accumulate the results into `grad` array without allocating a large temporary `gradient_per_sample` datastructure that does not fit in the CPU cache.

This could probably allow for significant speed-ups as observed in #21462 on machine with 16 cores and more.

But this is out of the scope of this PR. Just food for thoughts.